File: lib/solanum/source/network.rb
require 'solanum/source'

# bytes - The total number of bytes of data transmitted or received by the interface.
# packets - The total number of packets of data transmitted or received by the interface.
# errs - The total number of transmit or receive errors detected by the device driver.
# drop - The total number of packets dropped by the device driver.
# fifo - The number of FIFO buffer errors.
# frame - The number of packet framing errors.
# colls - The number of collisions detected on the interface.
# compressed - The number of compressed packets transmitted or received by the device driver. (This appears to be unused in the 2.2.15 kernel.)
# carrier - The number of carrier losses detected by the device driver.
# multicast - The number of multicast frames transmitted or received by the device driver.
class Solanum::Source::Network < Solanum::Source
  attr_reader :interfaces, :detailed

  STAT_FILE = '/proc/net/dev'

  FIELDS = %w{
    rx_bytes rx_packets rx_errs rx_drop rx_fifo rx_frame rx_compressed rx_multicast
    tx_bytes tx_packets tx_errs tx_drop tx_fifo tx_colls tx_carrier tx_compressed
  }

  SIMPLE_FIELDS = %w{rx_bytes rx_packets tx_bytes tx_packets}


  def initialize(opts)
    super(opts)
    @interfaces = opts['interfaces'] || []
    @detailed = opts['detailed'] || false
    @last = {}
  end


  def parse_stats(line)
    columns = line.strip.split(/\s+/)
    iface = columns.shift.chomp(':')
    return iface, Hash[FIELDS.zip(columns.map(&:to_i))]
  end


  def collect!
    events = []

    File.readlines(STAT_FILE).drop(2).each do |line|
      iface, stats = parse_stats(line)

      if @interfaces.empty? || @interfaces.include?(iface)
        if @last[iface]
          FIELDS.each do |field|
            next unless @detailed || SIMPLE_FIELDS.include?(field)
            diff = stats[field] - @last[iface][field]
            events << {
              service: "net #{iface} #{field.gsub('_', ' ')}",
              metric: diff,
            }
          end
        end
        @last[iface] = stats
      end
    end

    events
  end

end

File: src/solanum/source/network.clj
(ns solanum.source.network
  "Metrics source that measures network IO."
  (:require
    [clojure.string :as str]
    [solanum.source.core :as source]
    [solanum.system.linux :as linux]))


;; ## Measurements

(source/defsupport :network
  #{:linux})


(def ^:private simple-stats
  "Stats to show in simple mode."
  #{:rx-bytes :rx-packets :tx-bytes :tx-packets})


(def ^:private linux-net-fields
  "Fields in the net device file."
  [:rx-bytes :rx-packets :rx-errs :rx-drop :rx-fifo
   :rx-frame :rx-compressed :rx-multicast
   :tx-bytes :tx-packets :tx-errs :tx-drop :tx-fifo
   :tx-colls :tx-carrier :tx-compressed])


(defn- parse-net-counters
  "Parse a line from `/proc/net/dev` into a tuple with the interface name and
  a map of stat keys to numeric counter values."
  [line]
  (let [columns (str/split (str/trim line) #"\s+")]
    [(str/replace (first columns) #":$" "")
     (->> (rest columns)
          (map #(Long/parseLong %))
          (zipmap linux-net-fields))]))


(defn- read-net-data
  "Read network device counters from `/proc/net/dev`."
  []
  (into {}
        (comp
          (drop 2)
          (map parse-net-counters))
        (linux/read-proc-lines "/proc/net/dev")))


(defn- measure-linux-network
  "Measure network IO on Linux systems by reading the proc subsystem."
  [tracker]
  (let [data (read-net-data)
        prev @tracker]
    (reset! tracker data)
    (when prev
      (source/diff-tracker prev data))))


;; ## Network Source

(defrecord NetworkSource
  [tracker interfaces ignore detailed]

  source/Source

  (collect-events
    [this]
    (let [info (case (:mode this)
                 :linux (measure-linux-network tracker))]
      (into []
            (comp
              (if (seq interfaces)
                (filter (comp (set interfaces) key))
                identity)
              (if (seq ignore)
                (remove (comp (set ignore) key))
                identity)
              (mapcat
                (fn expand-events
                  [[iface diffs]]
                  (keep (fn net-event
                          [[stat diff]]
                          (when (or detailed (contains? simple-stats stat))
                            {:service (str "net io " (name stat))
                             :metric diff
                             :interface iface}))
                        diffs))))
            info))))


(defmethod source/initialize :network
  [config]
  (map->NetworkSource
    {:tracker (atom {})
     :interfaces (not-empty (set (:interfaces config)))
     :ignore (not-empty (set (:ignore config #{"lo"})))
     :detailed (boolean (:detailed config))}))
